Upgrade to 1.46 fails due to invalid state in permissions table

### Describe the bug

Previously we used to allow to have two paths that are identical except for the value after /schema/ in the permission table.

For example in the first path there's no value, and in the second path there is a value:

`/db/67/schema//table/283/query/segmented/`(no schema)
`/db/67/schema/dwh_name/table/283/query/segmented/`

Seems we used to allow this behaviour before 1.46.2 but not anymore since the upgrade fails

### To Reproduce

Not sure how to insert these kind of entries in the DB but there are people that have this issue. One such case in [stats](https://stats.metabase.com/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJxdWVyeSIsInF1ZXJ5Ijp7InNvdXJjZS10YWJsZSI6ODM0NCwiZmlsdGVyIjpbImFuZCIsWyI9IixbImZpZWxkIiwyODkxMjUsbnVsbF0sImFrdGl2byJdXSwiYWdncmVnYXRpb24iOltbImNvdW50Il1dLCJicmVha291dCI6W1siZmllbGQiLDI4ODYwNCxudWxsXV19LCJkYXRhYmFzZSI6NDV9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7InRhYmxlLmNvbHVtbl93aWR0aHMiOls1MzRdfSwib3JpZ2luYWxfY2FyZF9pZCI6bnVsbH0=):

When upgraded to 1.46.3 from 1.45.3.1 it started to fail with:

```
Reason: liquibase.exception.DatabaseException: ERROR: more than one row returned by a subquery used as an expression [Failed SQL: (0) UPDATE sandboxes s SET permission_id = (

Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set migrations/000_migrations.yaml::v46.00-066::noahmoss:
```

### Expected behavior

Upgrade works

### Logs

_No response_

### Information about your Metabase installation

```JSON
Migration from 1.45 to 1.46
```


### Severity

some customers are stuck in 1.45

### Additional context

_No response_

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14253,10 +14253,7 @@ databaseChangeLog:
             SET permission_id = (
               SELECT p.id
               FROM permissions p
-              JOIN (
-                SELECT id, db_id, schema
-                FROM metabase_table
-              ) t ON t.id = s.table_id
+              JOIN metabase_table t ON t.id = s.table_id
               WHERE
                 p.object LIKE CONCAT('/db/', t.db_id, '/schema/', t.schema, '/table/', s.table_id, '/query/segmented/')
                 AND p.group_id = s.group_id

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14255,9 +14255,9 @@ databaseChangeLog:
               FROM permissions p
               WHERE
                 p.object LIKE CONCAT('/db/',
-                                     (SELECT t.db_id FROM metabase_tables t WHERE t.id = s.table_id),
+                                     (SELECT t.db_id FROM metabase_table t WHERE t.id = s.table_id),
                                      '/schema/',
-                                     (SELECT t.schema FROM metabase_tables t WHERE t.id = s.table_id),
+                                     (SELECT t.schema FROM metabase_table t WHERE t.id = s.table_id),
                                      '/table/',
                                      s.table_id,
                                      '/query/segmented/')

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14254,13 +14254,13 @@ databaseChangeLog:
               SELECT p.id
               FROM permissions p
               WHERE
-                p.object LIKE CONCAT('/db/',
-                                     (SELECT t.db_id FROM metabase_table t WHERE t.id = s.table_id),
-                                     '/schema/',
-                                     (SELECT t.schema FROM metabase_table t WHERE t.id = s.table_id),
-                                     '/table/',
-                                     s.table_id,
-                                     '/query/segmented/')
+                p.object = CONCAT('/db/',
+                                  (SELECT t.db_id FROM metabase_table t WHERE t.id = s.table_id),
+                                  '/schema/',
+                                  (SELECT t.schema FROM metabase_table t WHERE t.id = s.table_id),
+                                  '/table/',
+                                  s.table_id,
+                                  '/query/segmented/')
                 AND p.group_id = s.group_id
               LIMIT 1
             )

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14255,9 +14255,9 @@ databaseChangeLog:
               FROM permissions p
               WHERE
                 p.object LIKE CONCAT('/db/',
-                                     (SELECT t.db_id FROM tables t WHERE t.id = s.table_id),
+                                     (SELECT t.db_id FROM metabase_tables t WHERE t.id = s.table_id),
                                      '/schema/',
-                                     (SELECT t.schema FROM tables t WHERE t.id = s.table_id),
+                                     (SELECT t.schema FROM metabase_tables t WHERE t.id = s.table_id),
                                      '/table/',
                                      s.table_id,
                                      '/query/segmented/')

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14245,21 +14245,21 @@ databaseChangeLog:
   - changeSet:
       id: v46.00-088
       author: noahmoss
-      comment: Added 0.46.5 -- backfill `permission_id` values in `sandboxes`. This is a fixed verison of v46.00-066
-               which has been removed, since it had a bug that blocked a customer from upgrading.
+      comment: Added 0.46.5 -- backfill `permission_id` values in `sandboxes`. This is a fixed version of v46.00-066
+               which has been removed since it had a bug that blocked a customer from upgrading.
       changes:
         - sql: >-
-            UPDATE sandboxes s
+            UPDATE sandboxes
             SET permission_id = (
-              SELECT p.id
-              FROM permissions p
-              JOIN metabase_table t ON t.id = s.table_id
+              SELECT permissions.id
+              FROM permissions
+              JOIN metabase_table ON metabase_table.id = sandboxes.table_id
               WHERE
-                p.object LIKE CONCAT('/db/', t.db_id, '/schema/', t.schema, '/table/', s.table_id, '/query/segmented/')
-                AND p.group_id = s.group_id
+                permissions.object LIKE CONCAT('/db/', metabase_table.db_id, '/schema/', metabase_table.schema, '/table/', sandboxes.table_id, '/query/segmented/')
+                AND permissions.group_id = sandboxes.group_id
               LIMIT 1
             )
-            WHERE permission_id is NULL;
+            WHERE permission_id IS NULL;
       rollback:
         # not required, new values added here do not need to be dropped
 

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14245,18 +14245,21 @@ databaseChangeLog:
   - changeSet:
       id: v46.00-088
       author: noahmoss
-      comment: Added 0.46.5 -- backfill `permission_id` values in `sandboxes`. This is a fixed version of v46.00-066
-               which has been removed since it had a bug that blocked a customer from upgrading.
+      comment: Added 0.46.5 -- backfill `permission_id` values in `sandboxes`. This is a fixed verison of v46.00-066
+               which has been removed, since it had a bug that blocked a customer from upgrading.
       changes:
         - sql: >-
-            UPDATE sandboxes
+            UPDATE sandboxes s
             SET permission_id = (
-              SELECT permissions.id
-              FROM permissions
-              JOIN metabase_table ON metabase_table.id = sandboxes.table_id
+              SELECT p.id
+              FROM permissions p
+              JOIN (
+                SELECT id, db_id, schema
+                FROM metabase_table
+              ) t ON t.id = s.table_id
               WHERE
-                permissions.object LIKE CONCAT('/db/', metabase_table.db_id, '/schema/', metabase_table.schema, '/table/', sandboxes.table_id, '/query/segmented/')
-                AND permissions.group_id = sandboxes.group_id
+                p.object LIKE CONCAT('/db/', t.db_id, '/schema/', t.schema, '/table/', s.table_id, '/query/segmented/')
+                AND p.group_id = s.group_id
               LIMIT 1
             )
             WHERE permission_id IS NULL;

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14158,33 +14158,6 @@ databaseChangeLog:
               remarks: The ID of the corresponding permissions path for this sandbox
 
   - changeSet:
-      id: v46.00-066
-      author: noahmoss
-      comment: Added 0.46.0 -- backfill `permission_id` values in `sandboxes`
-      changes:
-        - sql: >-
-            UPDATE sandboxes s
-            SET permission_id = (
-              SELECT id
-              FROM permissions p
-              WHERE
-                p.object LIKE CONCAT('%/table/', s.table_id, '/query/segmented/')
-                AND p.group_id = s.group_id
-            );
-      rollback:
-        # not required, new values added here are dropped in the rollback of v46.00-065
-
-  - changeSet:
-      id: v46.00-067
-      author: noahmoss
-      comment: Added 0.46.0 -- remove orphaned entries in `sandboxes`
-      changes:
-        - sql: >-
-            DELETE FROM sandboxes
-            WHERE permission_id IS NULL;
-      rollback: # not required, orphaned sandboxes should not be re-added to the DB
-
-  - changeSet:
       id: v46.00-070
       author: calherries
       comment: Added 0.46.0 - add entity_id column to action
@@ -14214,45 +14187,6 @@ databaseChangeLog:
             tableName: report_card
             columnName: updated_at
             newDataType: DATETIME
-
-  - changeSet:
-      id: v46.00-075
-      author: noahmoss
-      comment: Add foreign key constraint on sandboxes.permission_id
-      changes:
-        - addForeignKeyConstraint:
-            baseTableName: sandboxes
-            baseColumnNames: permission_id
-            referencedTableName: permissions
-            referencedColumnNames: id
-            constraintName: fk_sandboxes_ref_permissions
-            nullable: true
-      rollback:
-      # Not required because a rollback would drop the `permission_id` column anyway
-
-  - changeSet:
-      id: v46.00-076
-      author: noahmoss
-      comment: Remove foreign key constraint on sandboxes.permission_id
-      changes:
-        - dropForeignKeyConstraint:
-            baseTableName: sandboxes
-            constraintName: fk_sandboxes_ref_permissions
-      rollback:
-      # Not required because a rollback would drop the `permission_id` column anyway
-
-  - changeSet:
-      id: v46.00-077
-      author: noahmoss
-      comment: Add foreign key constraint on sandboxes.permission_id with the appropriate onDelete policy
-      changes:
-        - addForeignKeyConstraint:
-            baseTableName: sandboxes
-            baseColumnNames: permission_id
-            referencedTableName: permissions
-            referencedColumnNames: id
-            constraintName: fk_sandboxes_ref_permissions
-            onDelete: CASCADE
 
   - changeSet:
       id: v46.00-079
@@ -14307,6 +14241,50 @@ databaseChangeLog:
       changes:
         - customChange:
             class: "metabase.db.custom_migrations.DeleteAbandonmentEmailTask"
+
+  - changeSet:
+      id: v46.00-088
+      author: noahmoss
+      comment: Added 0.46.5 -- backfill `permission_id` values in `sandboxes`. This is a fixed verison of v46.00-066
+               which has been removed, since it had a bug that blocked a customer from upgrading.
+      changes:
+        - sql: >-
+            UPDATE sandboxes s
+            SET permission_id = (
+              SELECT p.id
+              FROM permissions p
+              JOIN metabase_table t ON t.id = s.table_id
+              WHERE
+                p.object LIKE CONCAT('/db/', t.db_id, '/schema/', t.schema, '/table/', s.table_id, '/query/segmented/')
+                AND p.group_id = s.group_id
+              LIMIT 1
+            )
+            WHERE permission_id is NULL;
+      rollback:
+        # not required, new values added here do not need to be dropped
+
+  - changeSet:
+      id: v46.00-089
+      author: noahmoss
+      comment: Added 0.46.5 -- remove orphaned entries in `sandboxes`
+      changes:
+        - sql: >-
+            DELETE FROM sandboxes
+            WHERE permission_id IS NULL;
+      rollback: # not required, orphaned sandboxes should not be re-added to the DB
+
+  - changeSet:
+      id: v46.00-090
+      author: noahmoss
+      comment: Add foreign key constraint on sandboxes.permission_id
+      changes:
+        - addForeignKeyConstraint:
+            baseTableName: sandboxes
+            baseColumnNames: permission_id
+            referencedTableName: permissions
+            referencedColumnNames: id
+            constraintName: fk_sandboxes_ref_permissions
+            onDelete: CASCADE
 
   - changeSet:
       id: v47.00-001

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14253,9 +14253,14 @@ databaseChangeLog:
             SET permission_id = (
               SELECT p.id
               FROM permissions p
-              JOIN metabase_table t ON t.id = s.table_id
               WHERE
-                p.object LIKE CONCAT('/db/', t.db_id, '/schema/', t.schema, '/table/', s.table_id, '/query/segmented/')
+                p.object LIKE CONCAT('/db/',
+                                     (SELECT t.db_id FROM tables t WHERE t.id = s.table_id),
+                                     '/schema/',
+                                     (SELECT t.schema FROM tables t WHERE t.id = s.table_id),
+                                     '/table/',
+                                     s.table_id,
+                                     '/query/segmented/')
                 AND p.group_id = s.group_id
               LIMIT 1
             )

--- a/resources/migrations/000_migrations.yaml
+++ b/resources/migrations/000_migrations.yaml
@@ -14277,6 +14277,8 @@ databaseChangeLog:
       id: v46.00-090
       author: noahmoss
       comment: Add foreign key constraint on sandboxes.permission_id
+      # This is a revised version of a prior migration (v46.00-075) so it may fail due to the FK constriant already existing
+      failOnError: false
       changes:
         - addForeignKeyConstraint:
             baseTableName: sandboxes

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -1113,3 +1113,38 @@
           (is (= cases rollbacked-to-18))
           (is (true? (custom-migrations-test/no-cards-are-overlap? rollbacked-to-18)))
           (is (true? (custom-migrations-test/no-cards-are-out-of-grid-and-has-size-0? rollbacked-to-18 18))))))))
+
+(deftest backfill-permission-id-test
+  (testing "Migrations v47.00-043: backfill `permission_id` FK on sandbox table"
+    (impl/test-migrations ["v47.00-043" "v47.00-045"] [migrate!]
+      (let [db-id    (first (t2/insert-returning-pks! (t2/table-name Database) {:name       "DB"
+                                                                                :engine     "h2"
+                                                                                :created_at :%now
+                                                                                :updated_at :%now
+                                                                                :details    "{}"}))
+            table-id (first (t2/insert-returning-pks! (t2/table-name Table) {:db_id      db-id
+                                                                             :schema     "SchemaName"
+                                                                             :name       "Table"
+                                                                             :created_at :%now
+                                                                             :updated_at :%now
+                                                                             :active     true}))
+            _        (t2/query-one {:insert-into :sandboxes
+                                    :values      [{:group_id             1
+                                                   :table_id             table-id
+                                                   :attribute_remappings "{\"foo\", 1}"}
+                                                  {:group_id             2
+                                                   :table_id             table-id
+                                                   :attribute_remappings "{\"foo\", 1}"}]})
+            perm-id  (first (t2/insert-returning-pks! (t2/table-name Permissions)
+                                                      [{:group_id 1
+                                                        :object   "/db/1/schema/SchemaName/table/1/query/segmented/"}
+                                                       {:group_id 1
+                                                        :object   "/db/1/schema//table/1/query/segmented/"}]))]
+        ;; Two rows are present in `sandboxes`
+        (is (= [{:id 1, :group_id 1, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}" :permission_id nil}
+                {:id 2, :group_id 2, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}" :permission_id nil}]
+               (mdb.query/query {:select [:*] :from [:sandboxes]})))
+        (migrate!)
+        ;; Only the sandbox with a corresponding `Permissions` row is present
+        (is (= [{:id 1, :group_id 1, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}", :permission_id perm-id}]
+               (mdb.query/query {:select [:*] :from [:sandboxes]})))))))

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -836,38 +836,6 @@
                    "F2 D1"}
                  (t2/select-fn-set :name Dimension {:order-by [[:id :asc]]}))))))))
 
-(deftest clean-up-gtap-table-test
-  (testing "Migrations v46.00-064 to v46.00-067: rename `group_table_access_policy` table, add `permission_id` FK,
-           and clean up orphaned rows"
-    (impl/test-migrations ["v46.00-064" "v46.00-067"] [migrate!]
-      (let [db-id    (first (t2/insert-returning-pks! (t2/table-name Database) {:name       "DB"
-                                                                                :engine     "h2"
-                                                                                :created_at :%now
-                                                                                :updated_at :%now
-                                                                                :details    "{}"}))
-            table-id (first (t2/insert-returning-pks! (t2/table-name Table) {:db_id      db-id
-                                                                             :name       "Table"
-                                                                             :created_at :%now
-                                                                             :updated_at :%now
-                                                                             :active     true}))
-            _        (t2/query-one {:insert-into :group_table_access_policy
-                                    :values      [{:group_id             1
-                                                   :table_id             table-id
-                                                   :attribute_remappings "{\"foo\", 1}"}
-                                                  {:group_id             2
-                                                   :table_id             table-id
-                                                   :attribute_remappings "{\"foo\", 1}"}]})
-            perm-id  (first (t2/insert-returning-pks! (t2/table-name Permissions) {:group_id 1
-                                                                                   :object   "/db/1/schema/PUBLIC/table/1/query/segmented/"}))]
-        ;; Two rows are present in `group_table_access_policy`
-        (is (= [{:id 1, :group_id 1, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}"}
-                {:id 2, :group_id 2, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}"}]
-               (mdb.query/query {:select [:*] :from [:group_table_access_policy]})))
-        (migrate!)
-        ;; Only the sandbox with a corresponding `Permissions` row is present, and the table is renamed to `sandboxes`
-        (is (= [{:id 1, :group_id 1, :table_id table-id, :card_id nil, :attribute_remappings "{\"foo\", 1}", :permission_id perm-id}]
-               (mdb.query/query {:select [:*] :from [:sandboxes]})))))))
-
 (deftest able-to-delete-db-with-actions-test
   (testing "Migrations v46.00-084 and v46.00-085 set delete CASCADE for action.model_id to
            fix the bug of unable to delete database with actions"
@@ -1115,8 +1083,8 @@
           (is (true? (custom-migrations-test/no-cards-are-out-of-grid-and-has-size-0? rollbacked-to-18 18))))))))
 
 (deftest backfill-permission-id-test
-  (testing "Migrations v47.00-043: backfill `permission_id` FK on sandbox table"
-    (impl/test-migrations ["v47.00-043" "v47.00-045"] [migrate!]
+  (testing "Migrations v46.00-088-v46.00-90: backfill `permission_id` FK on sandbox table"
+    (impl/test-migrations ["v46.00-088" "v46.00-090"] [migrate!]
       (let [db-id    (first (t2/insert-returning-pks! (t2/table-name Database) {:name       "DB"
                                                                                 :engine     "h2"
                                                                                 :created_at :%now

